Add flashcard tag search feature

The-Mad-Duck opened the pull request:
- Implement delete function
- Implement tag search with current flashcard tag feature

kwilson47 committed:
Added tags field to flashcard form.

kwilson47 committed:
Added functionality to retrieve existing tags for a flashcard and display them on the edit page.

kwilson47 committed:
Added ability to save tags to the database.

kwilson47 committed:
Added ability to filter cards based on tags on the student page.

kwilson47 committed:
Fixed bugs involving tags with spaces and carousel navigation.

kwilson47 committed:
Fixed bugs involving removing all tags from a card and viewing cards with no tags.

The-Mad-Duck committed:
Add tags feature to card search

kwilson47 committed:
Restored tag filter functionality. Now convert tags to uppercase when saving.

angrychai commented on src/Mentat.Repository/Models/CardDatabaseSettings.cs:
This class will need to be removed and the new database options for it implemented in appsecrets.json

dyohn replied:
This will get automatically resolved as long as the options branch and identity merge first to development. Then this branch can pull those changes from `origin/development` and be updated to start using the `appsecrets.json w/ options pattern` approach.

angrychai commented on src/Mentat.Repository/Services/CardService.cs:
Calls to the DB need exception handling added.

angrychai commented on src/Mentat.UI/appsettings.json:
Connection strings will need move to appsecrets.json

angrychai reviewed (dismissed):
Notes were added for some changes that will need to be made to interop with other branches.

Notes from my testing of this branch:
filtering of cards by difficulty works
Adding, editing, and deleting cards works
Could not manipulate tags on cards
Search box on filtering page may not be currently functioning.

dyohn commented on src/Mentat.Domain/Service/StudentService.cs:
```diff
@@ -19,12 +19,14 @@ ICardService cardService
                 FilterCount = selectedDifficulties?.Count ?? 0,
-                SelectedDifficulties = selectedDifficulties
+                SelectedDifficulties = selectedDifficulties,
+                AllUniqueTags = tags
```
Are tags guaranteed to be unique when they are read from the database?

kwilson47 replied:
I think so. tags is a List<string> that is populated by GetAllTags(). It gets all flashcard documents, loops through them one by one, and adds any tags to a HashSet in order to ensure uniqueness. Then it converts the HashSet to a List.
```csharp 
            HashSet<string> uniqueTags = new HashSet<string>();
            foreach (var doc in documents)
            {
                if (doc.Tags != null)
                    uniqueTags.UnionWith(doc.Tags);
            }
            return uniqueTags.ToList();
```

kwilson47 committed:
Various enhancements related to tags

Added a button to show all flashcards
Changed search to only search flashcards that are currently being filtered
Fixed bug in carousel that was causing two cards to display in certain situations
Removed unnecessary Tags header from Card Index page
Removed unnecessary hidden classes from carousel items

dyohn commented on src/Mentat.Repository/Models/CardDatabaseSettings.cs:
```diff
@@ -2,7 +2,7 @@
     public class CardDatabaseSettings : ICardDatabaseSettings
     {
-        public string ConnectionString { get; set; } = "mongodb+srv://cdb:eN9qAv7forAKSnEt@flashcardapp.4yggo3z.mongodb.net/?retryWrites=true&w=majority";
+        public string ConnectionString { get; set; } = "mongodb+srv://feature-tags-user:cFVOCEplar0AgjOq@cluster0.5mowjp1.mongodb.net/test";
```
This is going to be a problem. Not only do we have existing functionality that depends on this value, `CardDatabaseSettings` will be handled differently once `Identity` branch merges. I strongly suspect that once that branch and the Bash script generator branch merge, this branch will have one or more merge conflicts anyway. That's not an unusual situation. We will try to push through those other two PRs so that perhaps we can get these conflicts resolved before end of semester. If we don't, I'll just have the next class resolve any blockers for this PR.

dyohn commented on src/Mentat.UI/Views/Student/Index.cshtml:
It seems like someone's formatter is set differently than the project's. This will also be an issue, but one which is fairly easily fixed once the blocking changes are incorporated via rebase.

dyohn commented on src/Mentat.UI/appsettings.json:
This will have to be changed once the other branches merge. We're moving to use of an `appsecrets.json` file, and even if we weren't we need to consolidate all this work into a single database/user.

dyohn reviewed (changes requested):
Overall, very well done. The problems that I see are either formatting or, more importantly, that the database interaction details are OBE (overcome by events), making them incompatible with the direction we are going. Changes becoming OBE is not an unusual event, you will encounter this many times in your career. 

Once the other two PRs merge (soon), we will have a better idea of what needs to be done to adjust this PR. If we run out of time, don't worry: I will just have the next cohort of students take care of this PR. The work your team did is valuable and will not be discarded.

mercykaye committed:
Add a check for keycode 13 in Student.js and added Tags display name in Index.cshtml

kwilson47 committed:
Added try catch block around database call.
